io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -111,7 +112,14 @@ data Failure =
      | FailureSloppyShutdown [LabeledThread]
   deriving Show
 
-instance Exception Failure
+instance Exception Failure where
+    displayException (FailureException err) = displayException  err
+    displayException FailureDeadlock = "<<io-sim deadlock>>"
+    displayException (FailureSloppyShutdown threads) =
+      concat [ "<<sloppy shutdown: "
```

Comment:
Do we want to include "io-sim" in both exception messages (deadlock and sloppy shutdown), or neither?

Reply:
I prefer both.

Reply:
And what about `Failure err`, maybe `"<<io-sim failure: " ++ displayException err ++ ">>"`?

Reply:
I decided against it, that would not work well if `displayException err` contained new line characters.